File: valid-parentheses/valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        parentheses = []

        for c in s:
            if c == '(' or c == '[' or c == '{':
                parentheses.append(c)
            else:
                if parentheses:
                    currentBracket = parentheses.pop()
                    if c == ')' and currentBracket != '(':
                        return False
                    elif c == ']' and currentBracket != '[':
                        return False
                    elif c == '}' and currentBracket != '{':
                        return False
                else:
                    if c == ')' or c == ']' or c == '}':
                        return False

        if parentheses:
            return False
        return True

File: valid-parentheses/valid_parentheses.go
func isValid(s string) bool {
	var parentheses []rune
	var currentBracket rune

	for _, c := range s {
		if c == '(' || c == '[' || c == '{' {
			parentheses = append(parentheses, c)
		} else {
			if len(parentheses) > 0 {
                // pop operation
				currentBracket, parentheses = parentheses[len(parentheses)-1], parentheses[:len(parentheses)-1]
				if c == ')' && currentBracket != '(' {
					return false
				} else if c == ']' && currentBracket != '[' {
					return false
				} else if c == '}' && currentBracket != '{' {
					return false
				}
			} else {
				return false
			}
		}
	}

	if len(parentheses) != 0 {
		return false
	}
	return true
}
